pkg/deploy/submariner.go

```diff
@@ -0,0 +1,213 @@
+/*
+SPDX-License-Identifier: Apache-2.0
+
+Copyright Contributors to the Submariner project.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package deploy
+
+import (
+	"encoding/base64"
+	submariner "github.com/submariner-io/submariner-operator/api/submariner/v1alpha1"
+	"github.com/submariner-io/submariner-operator/internal/constants"
+	"github.com/submariner-io/submariner-operator/pkg/broker"
+	"github.com/submariner-io/submariner-operator/pkg/client"
+	"github.com/submariner-io/submariner-operator/pkg/discovery/globalnet"
+	"github.com/submariner-io/submariner-operator/pkg/reporter"
+	"github.com/submariner-io/submariner-operator/pkg/secret"
+	"github.com/submariner-io/submariner-operator/pkg/subctl/operator/servicediscoverycr"
+	"github.com/submariner-io/submariner-operator/pkg/subctl/operator/submarinercr"
+	v1 "k8s.io/api/core/v1"
+	"strings"
+)
+
+func Submariner(clientProducer client.Producer, jo WithJoinOptions, brokerInfo broker.Info, brokerSecret *v1.Secret, netconfig globalnet.Config, imageOverrides map[string]string, status reporter.Interface) error {
+
+	pskSecret, err := secret.Ensure(clientProducer.ForKubernetes(), constants.OperatorNamespace, brokerInfo.IPSecPSK)
+	if err != nil {
+		return status.Error(err, "Error creating PSK secret for cluster")
+	}
+
+	submarinerSpec, err := populateSubmarinerSpec(jo, brokerInfo, brokerSecret, pskSecret, netconfig, imageOverrides)
+	if err != nil {
+		return status.Error(err, "Error populating Submariner spec")
+	}
+
+	err = submarinercr.Ensure(clientProducer.ForOperator(), constants.OperatorNamespace, submarinerSpec)
+	if err != nil {
+		return status.Error(err, "Submariner deployment failed")
+	}
+	return nil
+}
+
+func ServiceDiscovery(clientProducer client.Producer, jo WithJoinOptions, brokerInfo broker.Info, brokerSecret *v1.Secret, imageOverrides map[string]string, status reporter.Interface) error {
```

Comment:
Please move this to a separate _service_discovery.go_ file. 

Also I think it would be cleaner to define a separate `ServiceDiscoveryOptions`.